### PR TITLE
⬆️🐍🐧 Upgrade to Python 3 / neurodebian:bionic

### PR DESCRIPTION
**Pull request by shnizzedy**
This PR is intended to

- upgrade all remaining Python 2 syntax to Python 3 
- upgrade Docker base image from neurodebian:xenial to neurodebian:bionic
- drop support for Python 2

without otherwise affecting the user experience or outputs.

As a part of this upgrade, the following package upgrades / downgrades are included (:unlock: indicating a version is not specified and is resolved at build time):

- asn1crypto (:unlock:) 0.22.0 → 1.2.0
- cffi (:unlock:) 1.10.0 → 1.13.0
- cryptography (:unlock:) 1.8.1 → 2.8
- cython 0.26 → :unlock: (0.29.15)
- docutils (:unlock:) 0.16 → 0.15.2
- idna (:unlock:) 2.6 → 2.8
- INDI-tools 0.0.6 → [3a7b805](https://github.com/FCP-INDI/INDI-Tools/commit/3a7b805)
- matplotlib 2.0.2 → 3.1.3
- mock (:unlock:) 3.0.5 → 4.0.1
- more-itertools (:unlock:) 5.0.0 → 8.2.0
- networkx 1.11 → 2.4
- nibabel 2.3.0 → 2.3.3
- numpy 1.13.0 → 1.16.4
- packaging (:unlock:) 16.8 → 20.1
- Pillow (:unlock:) 6.2.2 → 7.0.0
- pip 9.0.1 → :unlock: (20.0.2)
- pycparser (:unlock:) 2.18 → 2.19
- pyOpenSSL (:unlock:) 17.0.0 → 19.0.0
- pyparsing (:unlock:) 2.2.0 → 2.4.6
- pytest (:unlock:) 4.6.9 → 5.3.5
- PyYAML (:unlock:) 4.2b1 → 5.3
- scikit-learn 0.19.1 → 0.22.1
- setuptools (:unlock:) 36.4.0 → 45.2.0
- scipy 1.2.1 → 1.4.1
- six (:unlock:) 1.10.0  → 1.12.0
- tornado (:unlock:) 5.1.1 → 6.0.3
- tqdm (:unlock:) 4.42.1 → 4.36.1
- urllib3 (:unlock:) 1.24.3 → 1.24.2
- wxpython 3.0.0.0 → :unlock: (4.0.4)
- zipp (:unlock:) 1.1.0 → 3.0.0

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,14 +1,22 @@
...
+
 RUN apt-get update
 
 # Install the validator
-RUN apt-get install -y curl && \
-     curl -sL https://deb.nodesource.com/setup_11.x | bash - && \
-     apt-get install -y nodejs
-RUN npm install -g bids-validator
+RUN apt-get install -y apt-utils curl && \
+     curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.2/install.sh | bash
```
Comment (pn2200) [inline Dockerfile]: Can't the same effect be achieved by `curl https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.2/install.sh | bash`? `curl` sends its output to stdout by default.

File: Dockerfile [hunk 2]
```diff
@@ -43,9 +51,11 @@ RUN apt-get install -y \
       mesa-utils \
       netpbm \
       pkg-config \
+      rsync \
       tcsh \
       unzip \
       vim \
+      wget \
```
Comment (pn2200) [inline Dockerfile]: Seems unnecessary to install wget, if curl is already available.

File: Dockerfile [hunk 3]
```diff
@@ -62,6 +72,11 @@ RUN apt-get install -y \
       x11proto-print-dev \
       xutils-dev
 
+# Install libpng12
+RUN wget -q -O /tmp/libpng12.deb http://mirrors.kernel.org/ubuntu/pool/main/libp/libpng/libpng12-0_1.2.54-1ubuntu1_amd64.deb && \
```
Comment (pn2200) [inline Dockerfile]: Equivalent to `curl --silent --output /tmp/libpng12.deb http://mirrors.kernel.org/ubuntu/pool/main/libp/libpng/libpng12-0_1.2.54-1ubuntu1_amd64.deb`, I think, which means wget isn't needed.

File: dev/docker_data/bids_utils.py [hunk 4]
```diff
@@ -452,13 +457,17 @@ def bids_gen_cpac_sublist(bids_dir, paths_list, config_dict, creds_path, dbg=Fal
                     if "fMRI" in f_dict["acq"]:
                         if "fmap" not in subdict[f_dict["sub"]][f_dict["ses"]]:
                             subdict[f_dict["sub"]][f_dict["ses"]]["fmap"] = {}
-                        if "epi_{0}".format(pe_dir) not in subdict[f_dict["sub"]][f_dict["ses"]]["fmap"].keys():
+                        if "epi_{0}".format(
```
Comment (pn2200) [inline dev/docker_data/bids_utils.py]: Similar to a previous comment: Just ask if key in the dict as a whole, without calling the `keys()` method.

File: requirements.txt [hunk 5]
```diff
@@ -1,27 +1,26 @@
 boto3==1.7.37
 click==6.7
 configparser==3.7.4
-cython==0.26
+cython
 future==0.16.0
-INDI-Tools==0.0.6
+git+git://github.com/shnizzedy/INDI-Tools.git@3a7b80558bd866bf57e2d911f75aeb8bd62b2723
```
Comment (pn2200) [inline requirements.txt]: This version is different from that used in `info.py`. Any reason for the discrepancy?

File: CPAC/GUI/interface/pages/alff.py [hunk 6]
```diff
@@ -9,54 +9,54 @@
 class ALFF(wx.html.HtmlWindow):
 
     def __init__(self, parent, counter  = 0):
-        from urllib2 import urlopen
+        from urllib.request import urlopen
```
[the remark this answers is not in the document]
Comment (shnizzedy) [reply CPAC/GUI/interface/pages/alff.py]: Resolved in 1a40a47717bbf65c1c0ba84dad576374b686ebec

File: CPAC/GUI/interface/windows/main_window.py [hunk 7]
```diff
@@ -524,7 +524,7 @@ def OnShow(self, event):
                     dlg = wx.lib.dialogs.ScrolledMessageDialog(self, msg, name, size = (800,800))
                     dlg.ShowModal()
                 except:
-                    print "Cannot open file %s"%(name)
+                    print("Cannot open file %s"%(name))
```
[the remark this answers is not in the document]
Comment (shnizzedy) [reply CPAC/GUI/interface/windows/main_window.py]: Resolved in fe2cf57331dda0dfe672f5d3f90e74331e4291fb

File: CPAC/GUI/interface/windows/main_window.py [hunk 8]
```diff
@@ -540,17 +540,17 @@ def OnDisplay(self, event):
                     dlg = wx.lib.dialogs.ScrolledMessageDialog(self, msg, name, size = (800,1000))
                     dlg.ShowModal()
                 except:
-                    print "Cannot open file %s"%(name)
+                    print("Cannot open file %s"%(name))
```
[the remark this answers is not in the document]
Comment (shnizzedy) [reply CPAC/GUI/interface/windows/main_window.py]: Resolved in fe2cf57331dda0dfe672f5d3f90e74331e4291fb

File: CPAC/qc/utils.py [hunk 9]
```diff
@@ -225,7 +225,7 @@ def first_pass_organizing_files(qc_path):
 
             qc_filename = qc_filename + fwhm_val + hp_lp_
 
-        if strat_dict.keys() == []:
+        if list(strat_dict.keys()) == []:
```
[the remark this answers is not in the document]
Comment (shnizzedy) [reply CPAC/qc/utils.py]: Resolved in ce342cb

File: CPAC/reho/reho.py [hunk 10]
```diff
@@ -2,7 +2,7 @@
 import os
 import sys
 import re
-import commands
+import subprocess
```
[the remark this answers is not in the document]
Comment (shnizzedy) [reply CPAC/reho/reho.py]: Resolved in 1d24d1d

File: CPAC/utils/bids_utils.py [hunk 11]
```diff
@@ -461,15 +461,15 @@ def bids_gen_cpac_sublist(bids_dir, paths_list, config_dict, creds_path,
                         if "fmap" not in subdict[f_dict["sub"]][
                             f_dict["ses"]]:
                             subdict[f_dict["sub"]][f_dict["ses"]]["fmap"] = {}
-                        if "epi_{0}".format(pe_dir) not in \
-                                subdict[f_dict["sub"]][f_dict["ses"]][
-                                    "fmap"].keys():
+                        if "epi_{0}".format(pe_dir) not in subdict[
+                            f_dict["sub"]
+                        ][f_dict["ses"]]["fmap"].keys():
```
[the remark this answers is not in the document]
Comment (shnizzedy) [reply CPAC/utils/bids_utils.py]: Resolved in 7a9f223

File: CPAC/utils/extract_data_multiscan.py [hunk 12]
```diff
@@ -454,16 +454,16 @@ def read_csv(csv_input):
...
         dict_labels = defaultdict(list)
         for line in reader:
-            csv_dict = dict((k.lower(), v) for k, v in line.iteritems())
+            csv_dict = dict((k.lower(), v) for k, v in line.items())
             dict_labels[csv_dict.get('site'), csv_dict.get('scan')] = \
             [csv_dict[key] for key in sorted(csv_dict.keys()) \
              if key != 'site' and key != 'scan']
 
-        if len(dict_labels.keys()) < 1:
+        if len(list(dict_labels.keys())) < 1:
```
[the remark this answers is not in the document]
Comment (shnizzedy) [reply CPAC/utils/extract_data_multiscan.py]: Resolved in 28ccb8b

File: CPAC/vmhc/vmhc.py [hunk 13]
```diff
@@ -1,12 +1,12 @@
 import sys
 import os
-import commands
+import subprocess
```
[the remark this answers is not in the document]
Comment (shnizzedy) [reply CPAC/vmhc/vmhc.py]: Resolved in 1d24d1d

File: Dockerfile (hunk 1 above)
Comment (shnizzedy) [reply Dockerfile]: Resolved in b5b1714

File: Dockerfile (hunk 2 above)
Comment (shnizzedy) [reply Dockerfile]: Resolved in b5b1714

File: Dockerfile (hunk 3 above)
Comment (shnizzedy) [reply Dockerfile]: Resolved in b5b1714 + 5bd660d

File: dev/docker_data/bids_utils.py (hunk 4 above)
Comment (shnizzedy) [reply dev/docker_data/bids_utils.py]: Resolved in 72066d8

File: requirements.txt (hunk 5 above)
Comment (shnizzedy) [reply requirements.txt]: In this PR `info.py` does not specify a version: https://github.com/FCP-INDI/C-PAC/blob/e9b866eb22e9cd43b5d4959e6a8f20b0ffb2b9bb/CPAC/info.py#L140

INDI-Tools was [migrated to Python 3 in 2018](https://github.com/FCP-INDI/INDI-Tools/commit/f052440335913d03f89b75504bc9a6577877308b) but [the PyPI distribution](https://pypi.org/project/INDI-Tools/) hasn't been updated yet. I made [a fork](https://github.com/FCP-INDI/INDI-Tools/pull/1/files) to [increment the version and classifiers](https://github.com/FCP-INDI/INDI-Tools/pull/1/files) and [install from in the meantime](https://github.com/FCP-INDI/C-PAC/pull/1200/commits/e368ba780c8a4e02f4d81bf0810268f2c522f784) until we update the PyPI distro.

File: CPAC/utils/create_flame_model_files.py [hunk 14]
```diff
@@ -38,19 +38,19 @@ def write_mat_file(design_matrix, output_dir, model_name, \
 
     with open(out_file, 'wt') as f:
 
-        print >>f, '/NumWaves\t%d' %dimy
-        print >>f, '/NumPoints\t%d' %dimx
-        print >>f, ppstring
+        print('/NumWaves\t%d' %dimy, file=f)
```
[the remark this answers is not in the document]
Comment (shnizzedy) [reply CPAC/utils/create_flame_model_files.py]: I believe `print` puts a newline at the end of each call whereas `write` does not.

File: CPAC/pipeline/cpac_pipeline.py [hunk 15]
```diff
@@ -3382,8 +3383,8 @@ def prep_workflow(sub_dict, c, run, pipeline_timing_info=None,
 
 
         # Connectome
-        if "PearsonCorr" in ts_analysis_dict.keys() or \
-            "PartialCorr" in ts_analysis_dict.keys():
+        if "PearsonCorr" in ts_analysis_dict.keys(
+        ) or "PartialCorr" in ts_analysis_dict.keys():
```
[the remark this answers is not in the document]
Comment (shnizzedy) [reply CPAC/pipeline/cpac_pipeline.py]: Why? I favor breaking on parenthetical openers to breaking with a backslash, as recommended in [PEP 8: Maximum Line Length](https://www.python.org/dev/peps/pep-0008/#maximum-line-length):

> The preferred way of wrapping long lines is by using Python's implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.

**Comment by shnizzedy** [comment]
Thanks @pn2200! Several of you suggestions (particularly dropping the unused `urlopen` and `subprocess` imports) were things I was planning on doing once I got this working anyway. I went ahead and updated the PR with most of your suggestions, and left comments on the suggestions I did not apply.

**Comment by shnizzedy** [comment]
fcpindi/c-pac:python3 is up on Docker Hub, ready to test

File: CPAC/pipeline/cpac_pipeline.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (pn2200) [reply CPAC/pipeline/cpac_pipeline.py]: Okay, thanks for looking into it.

File: CPAC/utils/create_flame_model_files.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (pn2200) [reply CPAC/utils/create_flame_model_files.py]: That's right, so my suggestion was not entirely correct. Again, thanks for looking at it.

File: requirements.txt (hunk 5 above)
Comment (pn2200) [reply requirements.txt]: That makes sense to maintain the specific version used here until the package is updated in PyPI.

File: CPAC/qc/utils.py [hunk 16]
```diff
@@ -1840,9 +1843,9 @@ def make_montage_sagittal(overlay, underlay, png_name, cbar_name):
     matplotlib.rcParams.update({'font.size': 5})
 
     try:
-        from mpl_toolkits.axes_grid1 import ImageGrid   
+        from mpl_toolkits.axes_grid1 import ImageGrid
     except:
-        from mpl_toolkits.axes_grid import ImageGrid
+        from mpl_toolkits.axes_grid1 import ImageGrid
```
Comment (pn2200) [inline CPAC/qc/utils.py]: This `try-except` is not necessary; just import `from mpl_toolkits.axes_grid1` one time only.

Comment (shnizzedy) [reply CPAC/qc/utils.py]: Thanks for catching that @pn2200! Resolved in af3e712a

**Comment by sgiavasis** [comment]
fcpindi/c-pac:python3 worked for me, in terms of pipeline construction and running. I didn't see any impact on actual calculations and derivations with a smaller-scale set of correlations (and I doubt it would have impacted it), but we will find out for sure with the regression test.

**Pull request merged**